File: frontend/js/weather.js
/**
 * FarmIntel AI - Real-Time Weather Analysis
 * Fetches weather data from backend API
 */

// Initialize AOS
AOS.init({
    duration: 800,
    once: true,
    offset: 50
});

// Configuration
const API_CONFIG = {
    BASE_URL: 'http://localhost:5000/api', // Your Flask backend URL
    UPDATE_INTERVAL: 300000 // 5 minutes
};

// DOM Elements
const locationInput = document.getElementById('locationInput');
const searchLocationBtn = document.getElementById('searchLocationBtn');
const detectLocationBtn = document.getElementById('detectLocationBtn');
const refreshWeatherBtn = document.getElementById('refreshWeatherBtn');
const loadingSpinner = document.getElementById('loadingSpinner');
const weatherLocationSpan = document.getElementById('weatherLocation').querySelector('span');
const currentTemp = document.getElementById('currentTemp');
const weatherCondition = document.getElementById('weatherCondition');
const humidity = document.getElementById('humidity');
const windSpeed = document.getElementById('windSpeed');
const rainfall = document.getElementById('rainfall');
const minTemp = document.getElementById('minTemp');
const maxTemp = document.getElementById('maxTemp');
const pressure = document.getElementById('pressure');
const riskValue = document.getElementById('riskValue');
const riskFill = document.getElementById('riskFill');
const riskMessage = document.getElementById('riskMessage');
const riskBadge = document.getElementById('riskBadge');
const alertsContainer = document.getElementById('alertsContainer');
const sprayAdvisory = document.getElementById('sprayAdvisory');
const trendAnalysis = document.getElementById('trendAnalysis');
const updateTimeSpan = document.querySelector('#updateTime span');

let weatherChart = null;
let currentLocation = 'Chandigarh';
let updateTimer = null;

// Create particles
function createParticles() {
    const container = document.getElementById('particles');
    if (!container) return;
    
    for (let i = 0; i < 50; i++) {
        const p = document.createElement('div');
        p.classList.add('particle');
        p.style.width = Math.random() * 5 + 2 + 'px';
        p.style.height = p.style.width;
        p.style.left = Math.random() * 100 + '%';
        p.style.bottom = '-10px';
        p.style.animationDelay = Math.random() * 15 + 's';
        p.style.animationDuration = 8 + Math.random() * 12 + 's';
        p.style.background = `rgba(79, 70, 229, ${Math.random() * 0.3 + 0.1})`;
        container.appendChild(p);
    }
}

// Show toast notification
function showToast(message, type = 'info') {
    const container = document.getElementById('toastContainer');
    if (!container) return;
    
    const toast = document.createElement('div');
    toast.className = `toast ${type}`;
    const icons = {
        success: 'fa-check-circle',
        error: 'fa-exclamation-circle',
        info: 'fa-info-circle',
        warning: 'fa-exclamation-triangle'
    };
    toast.innerHTML = `<i class="fas ${icons[type] || icons.info}"></i><span>${message}</span>`;
    container.appendChild(toast);
    
    setTimeout(() => {
        if (toast.parentNode) toast.remove();
    }, 4000);
}

// Show/hide loading spinner
function setLoading(show) {
    if (loadingSpinner) {
        loadingSpinner.style.display = show ? 'block' : 'none';
    }
}

// Get weather icon class
function getWeatherIconClass(iconCode) {
    const iconMap = {
        '01d': 'fa-sun',
        '01n': 'fa-moon',
        '02d': 'fa-cloud-sun',
        '02n': 'fa-cloud-moon',
        '03d': 'fa-cloud',
        '03n': 'fa-cloud',
        '04d': 'fa-cloud',
        '04n': 'fa-cloud',
        '09d': 'fa-cloud-rain',
        '09n': 'fa-cloud-rain',
        '10d': 'fa-cloud-sun-rain',
        '10n': 'fa-cloud-moon-rain',
        '11d': 'fa-bolt',
        '11n': 'fa-bolt',
        '13d': 'fa-snowflake',
        '13n': 'fa-snowflake',
        '50d': 'fa-smog',
        '50n': 'fa-smog'
    };
    return iconMap[iconCode] || 'fa-cloud-sun-rain';
}

// Fetch weather data from backend API
async function fetchWeatherData(location) {
    try {
        setLoading(true);
        
        // Build query parameters
        const params = new URLSearchParams();
        if (location) {
            params.append('city', location);
        }
        
        const url = `${API_CONFIG.BASE_URL}/weather?${params.toString()}`;
        
        const response = await fetch(url);
        
        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            throw new Error(errorData.message || `API Error: ${response.status}`);
        }
        
        const data = await response.json();
        
        if (!data.success) {
            throw new Error(data.message || 'Failed to fetch weather data');
        }
        
        return data;
        
    } catch (error) {
        console.error('Weather API Error:', error);
        showToast(`Failed to fetch weather: ${error.message}`, 'error');
        return null;
    } finally {
        setLoading(false);
    }
}

// Fetch forecast from backend API
async function fetchForecast(location) {
    try {
        const params = new URLSearchParams();
        if (location) {
            params.append('city', location);
        }
        params.append('days', '7');
        
        const url = `${API_CONFIG.BASE_URL}/weather/forecast?${params.toString()}`;
        
        const response = await fetch(url);
        
        if (!response.ok) {
            return null;
        }
        
        const data = await response.json();
        return data.success ? data.forecast : null;
        
    } catch (error) {
        console.warn('Forecast API Error:', error);
        return null;
    }
}

// Update UI with weather data
async function updateWeatherData(location) {
    showToast(`Fetching real-time weather for ${location}...`, 'info');
    
    try {
        const data = await fetchWeatherData(location);
        
        if (!data) {
            return;
        }
        
        // Extract data from API response
        const weather = data.current || {};
        const locationInfo = data.location || {};
        const diseaseRisk = data.disease_risk || {};
        const alerts = data.alerts || [];
        const sprayAdvisoryData = data.spray_advisory || {};
        const forecastData = data.forecast || [];
        
        // Update location input
        const cityName = locationInfo.city || location || 'Unknown';
        locationInput.value = cityName;
        currentLocation = cityName;
        
        // Update current weather
        currentTemp.textContent = weather.temperature || '--';
        weatherCondition.textContent = weather.condition || 'Unknown';
        weatherLocationSpan.textContent = `${cityName.toUpperCase()}${locationInfo.country ? `, ${locationInfo.country}` : ''}`;
        humidity.textContent = weather.humidity || '--';
        windSpeed.textContent = weather.wind_speed || '--';
        rainfall.textContent = weather.rainfall || '0.0';
        minTemp.textContent = weather.min_temp || weather.temperature || '--';
        maxTemp.textContent = weather.max_temp || weather.temperature || '--';
        if (pressure) pressure.textContent = weather.pressure || '--';
        
        // Update update time
        if (updateTimeSpan) {
            updateTimeSpan.textContent = `Last updated: ${new Date().toLocaleString()}`;
        }
        
        // Update weather icon
        const weatherIcon = document.getElementById('weatherIcon');
        const iconClass = getWeatherIconClass(weather.weather_icon || '01d');
        weatherIcon.innerHTML = `<i class="fas ${iconClass}"></i>`;
        
        // Update risk index
        const riskLevel = diseaseRisk.risk_level || 'Low';
        const riskScore = diseaseRisk.risk_score || 0;
        const riskColor = diseaseRisk.risk_color || '#10B981';
        const riskMsg = diseaseRisk.risk_message || 'Loading weather data...';
        
        riskValue.textContent = riskLevel;
        riskValue.style.color = riskColor;
        
        const riskPercent = riskLevel === 'Critical' ? 95 : 
                           riskLevel === 'High' ? 80 : 
                           riskLevel === 'Medium' ? 55 : 25;
        riskFill.style.width = `${riskPercent}%`;
        riskFill.style.background = riskColor;
        riskMessage.textContent = riskMsg;
        
        // Update risk badge
        if (riskBadge) {
            riskBadge.className = `risk-badge ${riskLevel.toLowerCase()}`;
            riskBadge.textContent = riskLevel;
        }
        
        // Update risk factors
        const temp = weather.temperature || 25;
        const hum = weather.humidity || 65;
        const wind = weather.wind_speed || 10;
        const rain = weather.rainfall || 0;
        
        const tempPercent = Math.min(100, Math.max(0, Math.abs(temp - 25) * 6));
        const humidityPercent = Math.min(100, hum);
        const windPercent = Math.min(100, wind * 2);
        const rainPercent = Math.min(100, rain * 8);
        
        document.getElementById('tempFactor').style.width = `${tempPercent}%`;
        document.getElementById('humidityFactor').style.width = `${humidityPercent}%`;
        document.getElementById('windFactor').style.width = `${windPercent}%`;
        document.getElementById('rainFactor').style.width = `${rainPercent}%`;
        
        // Color factor bars
        document.getElementById('tempFactor').className = `factor-fill ${temp > 35 || temp < 0 ? 'high' : temp > 30 || temp < 10 ? 'medium' : 'low'}`;
        document.getElementById('humidityFactor').className = `factor-fill ${hum > 80 ? 'high' : hum > 65 ? 'medium' : 'low'}`;
        document.getElementById('windFactor').className = `factor-fill ${wind > 25 ? 'high' : wind > 15 ? 'medium' : 'low'}`;
        document.getElementById('rainFactor').className = `factor-fill ${rain > 15 ? 'high' : rain > 5 ? 'medium' : 'low'}`;
        
        // Update factor statuses
        const tempStatus = document.getElementById('tempStatus');
        const humidityStatus = document.getElementById('humidityStatus');
        const windStatus = document.getElementById('windStatus');
        const rainStatus = document.getElementById('rainStatus');
        
        if (tempStatus) {
            tempStatus.textContent = temp > 35 ? 'Extreme' : temp > 30 ? 'High' : temp < 0 ? 'Freezing' : temp < 5 ? 'Cold' : 'Normal';
            tempStatus.className = `factor-status ${temp > 35 || temp < 0 ? 'danger' : temp > 30 || temp < 5 ? 'warning' : 'success'}`;
        }
        if (humidityStatus) {
            humidityStatus.textContent = hum > 80 ? 'Critical' : hum > 65 ? 'High' : 'Normal';
            humidityStatus.className = `factor-status ${hum > 80 ? 'danger' : hum > 65 ? 'warning' : 'success'}`;
        }
        if (windStatus) {
            windStatus.textContent = wind > 25 ? 'High' : wind > 15 ? 'Moderate' : 'Low';
            windStatus.className = `factor-status ${wind > 25 ? 'danger' : wind > 15 ? 'warning' : 'success'}`;
        }
        if (rainStatus) {
            rainStatus.textContent = rain > 15 ? 'Heavy' : rain > 5 ? 'Moderate' : rain > 0 ? 'Light' : 'None';
        }
        
        // Update alerts
        updateAlerts(alerts);
        
        // Update spray advisory
        if (sprayAdvisoryData && sprayAdvisoryData.message) {
            sprayAdvisory.textContent = sprayAdvisoryData.message;
        } else {
            sprayAdvisory.textContent = 'No spray advisory available.';
        }
        
        // Update trend analysis
        if (forecastData && forecastData.length > 0) {
            const avgHumidity = forecastData.reduce((sum, d) => sum + (d.humidity || 0), 0) / forecastData.length;
            const highRiskDays = forecastData.filter(d => d.disease_risk === 'High' || d.disease_risk === 'Critical').length;
            const rainyDays = forecastData.filter(d => (d.rainfall || 0) > 5).length;
            
            let analysis = '';
            if (avgHumidity > 75) {
                analysis += `High humidity expected (${Math.round(avgHumidity)}%) over next 7 days. `;
            } else if (avgHumidity > 60) {
                analysis += `Moderate humidity levels expected (${Math.round(avgHumidity)}%). `;
            } else {
                analysis += `Low humidity conditions expected (${Math.round(avgHumidity)}%). `;
            }
            
            if (highRiskDays > 0) {
                analysis += `${highRiskDays} day(s) of high/critical disease risk. `;
            }
            
            if (rainyDays > 0) {
                analysis += `${rainyDays} day(s) with rainfall expected. Apply fungicide before rain.`;
            } else {
                analysis += 'No significant rainfall expected. Continue regular monitoring.';
            }
            
            trendAnalysis.textContent = analysis;
        } else {
            trendAnalysis.textContent = 'No forecast data available for trend analysis.';
        }
        
        // Update forecast grid
        if (forecastData && forecastData.length > 0) {
            updateForecastGrid(forecastData);
        }
        
        // Update chart
        if (forecastData && forecastData.length > 0) {
            updateWeatherChart(forecastData);
        }
        
        showToast(`Weather updated for ${cityName}!`, 'success');
        
    } catch (error) {
        console.error('Error updating weather:', error);
        showToast(`Error: ${error.message}`, 'error');
    }
}

// Update alerts in UI
function updateAlerts(alerts) {
    alertsContainer.innerHTML = '';
    
    if (!alerts || alerts.length === 0) {
        alertsContainer.innerHTML = `
            <div class="alert-card success">
                <i class="fas fa-check-circle"></i>
                <div class="alert-content">
                    <h4>✅ Good Conditions</h4>
                    <p>Current weather conditions are favorable. Continue regular monitoring.</p>
                </div>
            </div>
        `;
        return;
    }
    
    alerts.forEach(alert => {
        const severity = alert.severity || 'info';
        const typeClass = severity === 'critical' ? 'high' : 
                         severity === 'high' ? 'high' : 
                         severity === 'warning' ? 'warning' : 
                         severity === 'success' ? 'success' : 'info';
        
        alertsContainer.innerHTML += `
            <div class="alert-card ${typeClass}">
                <i class="fas ${alert.icon || 'fa-info-circle'}"></i>
                <div class="alert-content">
                    <h4>${alert.title || 'Alert'}</h4>
                    <p>${alert.message || 'Weather alert'}</p>
                </div>
            </div>
        `;
    });
}

// Update forecast grid
function updateForecastGrid(forecast) {
    const grid = document.getElementById('forecastGrid');
    if (!forecast || forecast.length === 0) {
        grid.innerHTML = '<p style="color: #94A3B8; grid-column: 1/-1; text-align: center;">No forecast data available</p>';
        return;
    }
    
    grid.innerHTML = forecast.map(day => {
        const riskClass = (day.disease_risk || 'Low').toLowerCase();
        const iconClass = getWeatherIconClass(day.weather_icon || '01d');
        const temp = day.temperature || '--';
        const hum = day.humidity || '--';
        const risk = day.disease_risk || 'Low';
        
        return `
            <div class="forecast-card ${riskClass}">
                <div class="forecast-day">${day.day || day.date || '--'}</div>
                <i class="fas ${iconClass}"></i>
                <div class="forecast-temp">${temp}°C</div>
                <div class="forecast-humidity">${hum}%</div>
                <div class="forecast-risk ${riskClass}">${risk}</div>
            </div>
        `;
    }).join('');
}

// Update weather chart
function updateWeatherChart(forecast) {
    const ctx = document.getElementById('weatherTrendChart').getContext('2d');
    
    if (weatherChart) weatherChart.destroy();
    
    const labels = forecast.map(d => d.day || d.date || '');
    const temps = forecast.map(d => d.temperature || 0);
    const hums = forecast.map(d => d.humidity || 0);
    
    weatherChart = new Chart(ctx, {
        type: 'line',
        data: {
            labels: labels,
            datasets: [
                {
                    label: 'Temperature (°C)',
                    data: temps,
                    borderColor: '#FBBF24',
                    backgroundColor: 'rgba(251, 191, 36, 0.1)',
                    borderWidth: 2,
                    fill: true,
                    tension: 0.4,
                    pointBackgroundColor: '#FBBF24',
                    pointBorderColor: '#fff',
                    pointRadius: 4,
                    pointHoverRadius: 6
                },
                {
                    label: 'Humidity (%)',
                    data: hums,
                    borderColor: '#4F46E5',
                    backgroundColor: 'rgba(79, 70, 229, 0.1)',
                    borderWidth: 2,
                    fill: true,
                    tension: 0.4,
                    pointBackgroundColor: '#4F46E5',
                    pointBorderColor: '#fff',
                    pointRadius: 4,
                    pointHoverRadius: 6
                }
            ]
        },
        options: {
            responsive: true,
            maintainAspectRatio: true,
            plugins: {
                legend: {
                    position: 'top',
                    labels: { color: '#94A3B8', font: { size: 12 } }
                },
                tooltip: { mode: 'index', intersect: false }
            },
            scales: {
                y: { 
                    beginAtZero: true, 
                    grid: { color: 'rgba(255,255,255,0.05)' }, 
                    ticks: { color: '#94A3B8' } 
                },
                x: { 
                    grid: { display: false }, 
                    ticks: { color: '#94A3B8' } 
                }
            }
        }
    });
}

// Detect user's location
async function detectUserLocation() {
    showToast('Detecting your location...', 'info');
    
    try {
        // Try IP-based geolocation first
        const ipResponse = await fetch('https://ipapi.co/json/');
        if (ipResponse.ok) {
            const ipData = await ipResponse.json();
            const city = ipData.city || ipData.region || ipData.country_name;
            
            if (city && city !== 'Unknown') {
                showToast(`Location detected: ${city}!`, 'success');
                locationInput.value = city;
                currentLocation = city;
                await updateWeatherData(city);
                return;
            }
        }
    } catch (error) {
        console.warn('IP geolocation failed:', error);
    }
    
    // Try browser geolocation
    if (navigator.geolocation) {
        navigator.geolocation.getCurrentPosition(
            async (position) => {
                const { latitude, longitude } = position.coords;
                showToast(`Location detected! Fetching weather...`, 'success');
                
                try {
                    // Use the API to get weather by coordinates
                    const params = new URLSearchParams();
                    params.append('lat', latitude);
                    params.append('lon', longitude);
                    
                    const url = `${API_CONFIG.BASE_URL}/weather?${params.toString()}`;
                    const response = await fetch(url);
                    
                    if (response.ok) {
                        const data = await response.json();
                        const city = data.location?.city || 'Unknown';
                        locationInput.value = city;
                        currentLocation = city;
                        await updateWeatherData(city);
                    } else {
                        // Use coordinates as fallback
                        const coords = `${latitude.toFixed(4)},${longitude.toFixed(4)}`;
                        locationInput.value = coords;
                        currentLocation = coords;
                        await updateWeatherData(coords);
                    }
                } catch (error) {
                    showToast('Error with location service. Please enter city manually.', 'error');
                }
            },
            (error) => {
                let errorMessage = 'Unable to detect location. ';
                switch(error.code) {
                    case error.PERMISSION_DENIED:
                        errorMessage += 'Please allow location access or enter city manually.';
                        break;
                    case error.POSITION_UNAVAILABLE:
                        errorMessage += 'Location information unavailable.';
                        break;
                    case error.TIMEOUT:
                        errorMessage += 'Location request timed out.';
                        break;
                    default:
                        errorMessage += 'Please enter city manually.';
                }
                showToast(errorMessage, 'warning');
            }
        );
    } else {
        showToast('Geolocation not supported. Please enter city manually.', 'warning');
    }
}

// Search location by city
async function searchLocation() {
    const location = locationInput.value.trim();
    if (!location) {
        showToast('Please enter a city name', 'error');
        return;
    }
    
    currentLocation = location;
    await updateWeatherData(location);
}

// Refresh weather data
async function refreshWeather() {
    if (refreshWeatherBtn) {
        refreshWeatherBtn.classList.add('spinning');
    }
    await updateWeatherData(currentLocation);
    if (refreshWeatherBtn) {
        setTimeout(() => {
            refreshWeatherBtn.classList.remove('spinning');
        }, 1000);
    }
}

// Event Listeners
if (searchLocationBtn) searchLocationBtn.addEventListener('click', searchLocation);
if (detectLocationBtn) detectLocationBtn.addEventListener('click', detectUserLocation);
if (refreshWeatherBtn) refreshWeatherBtn.addEventListener('click', refreshWeather);
if (locationInput) locationInput.addEventListener('keypress', (e) => {
    if (e.key === 'Enter') searchLocation();
});

// Mobile sidebar
const mobileToggle = document.getElementById('mobileMenuToggle');
const sidebar = document.getElementById('sidebar');
const overlay = document.getElementById('sidebarOverlay');

if (mobileToggle) {
    mobileToggle.addEventListener('click', () => {
        sidebar.classList.toggle('mobile-open');
        overlay.classList.toggle('active');
    });
}
if (overlay) {
    overlay.addEventListener('click', () => {
        sidebar.classList.remove('mobile-open');
        overlay.classList.remove('active');
    });
}

// Logout
const logoutBtn = document.getElementById('logoutBtn');
if (logoutBtn) {
    logoutBtn.addEventListener('click', () => {
        localStorage.removeItem('isLoggedIn');
        localStorage.removeItem('userName');
        window.location.href = '../index.html';
    });
}

// Set user name
const userName = localStorage.getItem('userName') || 'John';
const userNameElement = document.getElementById('userName');
if (userNameElement) userNameElement.innerHTML = userName;

// Theme toggle
let isDark = false;
const themeToggle = document.getElementById('themeToggle');
if (themeToggle) {
    themeToggle.addEventListener('click', () => {
        isDark = !isDark;
        document.body.classList.toggle('dark-mode', isDark);
        themeToggle.innerHTML = isDark ? '<i class="fas fa-moon"></i>' : '<i class="fas fa-sun"></i>';
        localStorage.setItem('theme', isDark ? 'dark' : 'light');
        showToast(`${isDark ? 'Dark' : 'Light'} mode activated`, 'success');
    });
}

const savedTheme = localStorage.getItem('theme');
if (savedTheme === 'dark') {
    isDark = true;
    document.body.classList.add('dark-mode');
    if (themeToggle) themeToggle.innerHTML = '<i class="fas fa-moon"></i>';
}

// Initialize
createParticles();

// Start with default location
updateWeatherData('Chandigarh');

// Auto-refresh every 5 minutes
if (updateTimer) clearInterval(updateTimer);
updateTimer = setInterval(() => {
    if (currentLocation) {
        updateWeatherData(currentLocation);
    }
}, API_CONFIG.UPDATE_INTERVAL);

File: backend/app/routes/weather.py
"""
Weather Routes - Weather data and disease risk forecasting endpoints
Handles current weather, forecasts, and disease risk calculations
"""

from flask import Blueprint, request, jsonify, current_app
from datetime import datetime

from app.controllers.weather_controller import WeatherController
from app.models.schemas import WeatherRequest, ValidationError

weather_bp = Blueprint('weather', __name__)


@weather_bp.route('/weather', methods=['GET'])
def get_weather():
    """
    Get current weather data for a location
    GET /api/weather?city=Chandigarh
    GET /api/weather?lat=30.7333&lon=76.7794
    GET /api/weather?pincode=160001
    
    Query Parameters:
        city: City name (e.g., Chandigarh, Mumbai, Delhi, Bangalore)
        lat: Latitude
        lon: Longitude
        pincode: Indian postal code
        
    Returns:
        JSON with weather data and disease risk analysis
    """
    try:
        # Get query parameters
        city = request.args.get('city')
        lat = request.args.get('lat')
        lon = request.args.get('lon')
        pincode = request.args.get('pincode')
        
        # Convert lat/lon to float if provided
        if lat:
            lat = float(lat)
        if lon:
            lon = float(lon)
        
        # Validate request
        weather_request = WeatherRequest(city=city, lat=lat, lon=lon, pincode=pincode)
        weather_request.validate()
        
        # Get weather data
        if pincode:
            weather_data = WeatherController.get_weather_by_pincode(pincode)
        else:
            weather_data = WeatherController.get_weather(city=city, lat=lat, lon=lon)
        
        # Get air quality if coordinates available
        air_quality = None
        if lat and lon:
            air_quality = WeatherController.get_air_quality(lat, lon)
        
        return jsonify({
            'success': True,
            'location': {
                'city': weather_data.get('city'),
                'country': weather_data.get('country'),
                'coordinates': {
                    'lat': lat,
                    'lon': lon
                } if lat and lon else None
            },
            'current': {
                'temperature': weather_data.get('temperature'),
                'feels_like': weather_data.get('feels_like'),
                'humidity': weather_data.get('humidity'),
                'pressure': weather_data.get('pressure'),
                'wind_speed': weather_data.get('wind_speed'),
                'wind_direction': weather_data.get('wind_direction'),
                'rainfall': weather_data.get('rainfall'),
                'clouds': weather_data.get('clouds'),
                'condition': weather_data.get('condition'),
                'condition_main': weather_data.get('condition_main'),
                'weather_icon': weather_data.get('weather_icon'),
                'sunrise': weather_data.get('sunrise'),
                'sunset': weather_data.get('sunset')
            },
            'disease_risk': {
                'risk_level': weather_data.get('disease_risk'),
                'risk_score': weather_data.get('risk_score'),
                'risk_color': weather_data.get('risk_color'),
                'risk_message': weather_data.get('risk_message'),
                'disease_specific_risks': weather_data.get('disease_risks', [])
            },
            'alerts': weather_data.get('alerts', []),
            'spray_advisory': weather_data.get('spray_advisory'),
            'forecast': weather_data.get('forecast', []),
            'air_quality': air_quality,
            'is_mock': weather_data.get('is_mock', False),
            'timestamp': datetime.now().isoformat()
        })
        
    except ValidationError as e:
        return jsonify({
            'success': False,
            'error': 'Validation Error',
            'message': e.message,
            'field': e.field
        }), 400
    except Exception as e:
        current_app.logger.error(f"Weather API error: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e),
            'message': 'An error occurred while fetching weather data'
        }), 500


@weather_bp.route('/weather/forecast', methods=['GET'])
def get_weather_forecast():
    """
    Get 7-day weather forecast
    GET /api/weather/forecast?city=Chandigarh
    GET /api/weather/forecast?lat=30.7333&lon=76.7794
    
    Query Parameters:
        city: City name
        lat: Latitude
        lon: Longitude
        days: Number of days (1-7, default 7)
        
    Returns:
        JSON with forecast data
    """
    try:
        city = request.args.get('city')
        lat = request.args.get('lat')
        lon = request.args.get('lon')
        days = request.args.get('days', 7, type=int)
        
        # Limit days to 7
        days = min(days, 7)
        
        # Get forecast
        if city:
            weather_data = WeatherController.get_weather(city=city)
            forecast = weather_data.get('forecast', [])
        elif lat and lon:
            weather_data = WeatherController.get_weather(lat=float(lat), lon=float(lon))
            forecast = weather_data.get('forecast', [])
        else:
            weather_data = WeatherController.get_weather()
            forecast = weather_data.get('forecast', [])
        
        # Limit to requested days
        forecast = forecast[:days]
        
        return jsonify({
            'success': True,
            'location': city or WeatherController._get_city_from_config(),
            'days': len(forecast),
            'forecast': forecast,
            'timestamp': datetime.now().isoformat()
        })
        
    except Exception as e:
        current_app.logger.error(f"Forecast error: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e),
            'message': 'An error occurred while fetching forecast data'
        }), 500


@weather_bp.route('/weather/risk', methods=['GET'])
def get_disease_risk():
    """
    Calculate disease risk based on weather conditions
    GET /api/weather/risk?city=Chandigarh
    GET /api/weather/risk?temp=28&humidity=85&wind=12&rain=2.4
    
    Query Parameters:
        city: City name (to get real weather)
        temp: Temperature (if not using city)
        humidity: Humidity (if not using city)
        wind: Wind speed (if not using city)
        rain: Rainfall (if not using city)
        
    Returns:
        JSON with disease risk analysis
    """
    try:
        # Check if custom parameters provided
        temp = request.args.get('temp', type=float)
        humidity = request.args.get('humidity', type=float)
        wind = request.args.get('wind', type=float)
        rain = request.args.get('rain', type=float)
        
        if temp is not None and humidity is not None:
            # Calculate risk from provided parameters
            from app.config.weather_config import WeatherConfig
            
            risk = WeatherConfig.calculate_disease_risk(temp, humidity, wind or 0, rain or 0)
            spray_advisory = WeatherConfig.get_spray_advisory(temp, humidity, wind or 0, rain or 0)
            alerts = WeatherConfig.get_alert_messages({
                'temperature': temp,
                'humidity': humidity,
                'wind_speed': wind or 0,
                'rainfall': rain or 0
            })
            
            return jsonify({
                'success': True,
                'conditions': {
                    'temperature': temp,
                    'humidity': humidity,
                    'wind_speed': wind,
                    'rainfall': rain
                },
                'disease_risk': {
                    'risk_level': risk['risk'],
                    'risk_score': risk['score'],
                    'risk_color': risk['color'],
                    'risk_message': risk['message']
                },
                'alerts': alerts,
                'spray_advisory': spray_advisory,
                'timestamp': datetime.now().isoformat()
            })
        else:
            # Get risk from city weather
            city = request.args.get('city', WeatherConfig.DEFAULT_CITY)
            weather_data = WeatherController.get_weather(city=city)
            
            return jsonify({
                'success': True,
                'location': city,
                'current_conditions': {
                    'temperature': weather_data.get('temperature'),
                    'humidity': weather_data.get('humidity'),
                    'wind_speed': weather_data.get('wind_speed'),
                    'rainfall': weather_data.get('rainfall')
                },
                'disease_risk': {
                    'risk_level': weather_data.get('disease_risk'),
                    'risk_score': weather_data.get('risk_score'),
                    'risk_color': weather_data.get('risk_color'),
                    'risk_message': weather_data.get('risk_message'),
                    'disease_specific': weather_data.get('disease_risks', [])
                },
                'alerts': weather_data.get('alerts', []),
                'spray_advisory': weather_data.get('spray_advisory'),
                'timestamp': datetime.now().isoformat()
            })
        
    except Exception as e:
        current_app.logger.error(f"Disease risk error: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e),
            'message': 'An error occurred while calculating disease risk'
        }), 500


@weather_bp.route('/weather/alerts', methods=['GET'])
def get_weather_alerts():
    """
    Get weather alerts for a location
    GET /api/weather/alerts?city=Chandigarh
    
    Query Parameters:
        city: City name
        
    Returns:
        JSON with weather alerts
    """
    try:
        from app.config.weather_config import WeatherConfig
        city = request.args.get('city', WeatherConfig.DEFAULT_CITY)
        weather_data = WeatherController.get_weather(city=city)
        
        return jsonify({
            'success': True,
            'location': city,
            'alerts': weather_data.get('alerts', []),
            'total_alerts': len(weather_data.get('alerts', [])),
            'high_risk': weather_data.get('disease_risk') == 'HIGH',
            'timestamp': datetime.now().isoformat()
        })
        
    except Exception as e:
        current_app.logger.error(f"Weather alerts error: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e),
            'message': 'An error occurred while fetching weather alerts'
        }), 500


@weather_bp.route('/weather/spray-advisory', methods=['GET'])
def get_spray_advisory():
    """
    Get spray advisory based on weather conditions
    GET /api/weather/spray-advisory?city=Chandigarh
    GET /api/weather/spray-advisory?temp=28&humidity=85&wind=12
    
    Query Parameters:
        city: City name (to get real weather)
        temp: Temperature (if not using city)
        humidity: Humidity (if not using city)
        wind: Wind speed (if not using city)
        
    Returns:
        JSON with spray advisory
    """
    try:
        from app.config.weather_config import WeatherConfig
        
        temp = request.args.get('temp', type=float)
        humidity = request.args.get('humidity', type=float)
        wind = request.args.get('wind', type=float)
        
        if temp is not None and humidity is not None:
            advisory = WeatherConfig.get_spray_advisory(temp, humidity, wind or 0, 0)
            
            return jsonify({
                'success': True,
                'conditions': {
                    'temperature': temp,
                    'humidity': humidity,
                    'wind_speed': wind
                },
                'advisory': advisory,
                'recommendation': advisory.get('recommendation'),
                'timestamp': datetime.now().isoformat()
            })
        else:
            city = request.args.get('city', WeatherConfig.DEFAULT_CITY)
            weather_data = WeatherController.get_weather(city=city)
            
            return jsonify({
                'success': True,
                'location': city,
                'current_conditions': {
                    'temperature': weather_data.get('temperature'),
                    'humidity': weather_data.get('humidity'),
                    'wind_speed': weather_data.get('wind_speed')
                },
                'advisory': weather_data.get('spray_advisory'),
                'timestamp': datetime.now().isoformat()
            })
        
    except Exception as e:
        current_app.logger.error(f"Spray advisory error: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e),
            'message': 'An error occurred while getting spray advisory'
        }), 500


@weather_bp.route('/weather/cities', methods=['GET'])
def get_supported_cities():
    """
    Get list of supported cities
    GET /api/weather/cities
    
    Returns:
        JSON with list of supported cities
    """
    from app.config.weather_config import WeatherConfig
    
    return jsonify({
        'success': True,
        'total': len(WeatherConfig.SUPPORTED_CITIES),
        'cities': WeatherConfig.SUPPORTED_CITIES,
        'default_city': WeatherConfig.DEFAULT_CITY,
        'timestamp': datetime.now().isoformat()
    })


@weather_bp.route('/weather/health', methods=['GET'])
def weather_health():
    """
    Check weather service health
    GET /api/weather/health
    
    Returns:
        JSON with service status
    """
    from app.config.weather_config import WeatherConfig
    
    return jsonify({
        'success': True,
        'service': 'weather_api',
        'api_configured': WeatherConfig.is_api_configured(),
        'default_city': WeatherConfig.DEFAULT_CITY,
        'supported_cities': len(WeatherConfig.SUPPORTED_CITIES),
        'status': 'healthy',
        'timestamp': datetime.now().isoformat()
    })


# Helper method to get city from config
@staticmethod
def _get_city_from_config():
    """Get default city from configuration"""
    from app.config.weather_config import WeatherConfig
    return WeatherConfig.DEFAULT_CITY